File: src/platform_context_graph/parsers/languages/runtime_dependencies.py
"""Helpers for extracting runtime service dependencies from source files."""

from __future__ import annotations

import re

_SERVICES_ARRAY_RE = re.compile(r"services\s*:\s*\[(?P<body>.*?)\]", re.DOTALL)
_STRING_LITERAL_RE = re.compile(r"""(?P<quote>['"`])(?P<value>.*?)(?P=quote)""")
_SERVICE_NAME_RE = re.compile(r"^[a-z0-9][a-z0-9-]*$")


def _normalize_dependency_name(value: str, *, workload_name: str) -> str | None:
    """Normalize one declared service string into a canonical workload name."""
    candidate = value.strip()
    if not candidate or "${" in candidate:
        return None
    if candidate.startswith("/api/"):
        candidate = candidate[len("/api/") :]
    if candidate == workload_name:
        return None
    if "/" in candidate:
        return None
    if candidate in {"aws", "elastic", "elasticache"}:
        return None
    if not _SERVICE_NAME_RE.match(candidate):
        return None
    return candidate


def extract_runtime_service_dependencies(
    source: str,
    *,
    workload_name: str,
) -> list[str]:
    """Extract workload-style service dependencies from a runtime config block."""
    matches: list[str] = []
    seen: set[str] = set()
    for services_match in _SERVICES_ARRAY_RE.finditer(source):
        body = services_match.group("body")
        for token_match in _STRING_LITERAL_RE.finditer(body):
            dependency_name = _normalize_dependency_name(
                token_match.group("value"),
                workload_name=workload_name,
            )
            if dependency_name is None or dependency_name in seen:
                continue
            seen.add(dependency_name)
            matches.append(dependency_name)
    return matches


__all__ = ["extract_runtime_service_dependencies"]

File: go/internal/parser/runtime_dependencies.go
package parser

import "strings"

func extractRuntimeServiceDependencies(source string, workloadName string) []string {
	matches := make([]string, 0)
	seen := make(map[string]struct{})
	remaining := source

	for {
		index := strings.Index(remaining, "services")
		if index < 0 {
			break
		}
		remaining = remaining[index+len("services"):]
		colonIndex := strings.Index(remaining, ":")
		if colonIndex < 0 {
			continue
		}
		remaining = remaining[colonIndex+1:]
		openIndex := strings.Index(remaining, "[")
		if openIndex < 0 {
			continue
		}
		remaining = remaining[openIndex+1:]
		closeIndex := strings.Index(remaining, "]")
		if closeIndex < 0 {
			break
		}

		body := remaining[:closeIndex]
		remaining = remaining[closeIndex+1:]
		for _, value := range extractQuotedStrings(body) {
			dependencyName := normalizeRuntimeDependencyName(value, workloadName)
			if dependencyName == "" {
				continue
			}
			if _, ok := seen[dependencyName]; ok {
				continue
			}
			seen[dependencyName] = struct{}{}
			matches = append(matches, dependencyName)
		}
	}

	return matches
}

func extractQuotedStrings(body string) []string {
	values := make([]string, 0)
	var (
		quote  rune
		buffer strings.Builder
	)

	for _, r := range body {
		switch {
		case quote == 0 && (r == '\'' || r == '"' || r == '`'):
			quote = r
			buffer.Reset()
		case quote != 0 && r == quote:
			values = append(values, buffer.String())
			quote = 0
		case quote != 0:
			if r == '\\' {
				continue
			}
			buffer.WriteRune(r)
		}
	}

	return values
}

func normalizeRuntimeDependencyName(value string, workloadName string) string {
	candidate := strings.TrimSpace(value)
	if candidate == "" || strings.Contains(candidate, "${") {
		return ""
	}
	candidate = strings.TrimPrefix(candidate, "/api/")
	if candidate == workloadName {
		return ""
	}
	if strings.Contains(candidate, "/") {
		return ""
	}
	switch candidate {
	case "aws", "elastic", "elasticache":
		return ""
	}
	if !isRuntimeServiceName(candidate) {
		return ""
	}
	return candidate
}

func isRuntimeServiceName(value string) bool {
	if value == "" {
		return false
	}
	if value[0] < 'a' || value[0] > 'z' {
		return false
	}
	for _, r := range value[1:] {
		if (r < 'a' || r > 'z') && (r < '0' || r > '9') && r != '-' {
			return false
		}
	}
	return true
}
